explainer.md

```diff
@@ -59,12 +59,12 @@ Alternatively, the depth data is also available via the `depthInfo.data` attribu
 
 For example, to access the data at row `r`, column `c` of the buffer that has `"luminance-alpha"` format, the app can use:
 ```js
-const uint16Data = new Uint16Array(depthInfo.data.buffer,
-                                   depthInfo.data.byteOffset,
-                                   depthInfo.data.byteLength);
+const uint8Data = new Uint8Array(depthInfo.data.buffer,
```

Comment:
I'm not sure if I understand the reason for this change - LUMINANCE_ALPHA is [documented](https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/texImage2D) as 2-byte-per-pixel format (with 2 channels), and we definitely send out 2 bytes per pixel in our Chrome implementation (that represent 16-bit unsigned integer - the value needs to be reassembled on the GPU by inspecting the L & A channels, but `Uint16Array` should take care of that on the CPU).

Reply:
LUMINANCE_ALPHA is using UNSIGNED_BYTE data type in WebGL, so it should be created using Uint8Array. Please refer to this documentation that describes on that: https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/texImage2D

Before recent change on specs, Uint8Array was used, and it worked well. With Uint16Array, it does not work - array fails to be created due to incorrect arguments. in Chrome Canary 91, changing it to Uint8Array - works.

Reply:
Also, on slightly different note, with recent changes, things are still wrong, not sure it is a work in progress, or more complete changes, but data constructed from CPU path raw data to texture, is flipped vertically.

Reply:
> LUMINANCE_ALPHA is using UNSIGNED_BYTE data type in WebGL (...)

That's correct, but this section of the explainer describes how to access depth values on the CPU using the raw buffer (instead of relying on `depthInfo.getDepthInMeters()`), not how to upload the data to the texture.

> (...) data constructed from CPU path raw data to texture, is flipped vertically.

Can you verify if you've applied `normDepthBufferFromNormView` when sampling from the texture? Can you check if things work for you on our test [page](https://storage.googleapis.com/chromium-webxr-test/r853318/proposals/phone-ar-depth-gpu.html) (note that [texture upload](https://source.chromium.org/chromium/chromium/src/+/master:third_party/webxr_test_pages/webxr-samples/proposals/phone-ar-depth-gpu.html;l=366;drc=26f291b695a915c2ba97904c460034ce0da5e1d2) uses `Uint8Array` because WebGL1 mandates that)?

Reply:
> Can you verify if you've applied normDepthBufferFromNormView when sampling from the texture?

Yes, managed to narrow down why it is changed since previous version. In test test code provided, hard coded texture UVs are used:

```js
const vertices_data = [
  -1, -1,    0, 1,  // bottom left
   1, -1,    1, 1,  // bottom right
  -1,  1,    0, 0,  // top left
   1,  1,    1, 0,  // top right
];
```

With a related comment:
```js
// clip space coordinates + texture space coordinates
// our depth buffer has an origin in top-left corner of the screen -
// we need to adjust the texture coordinates to account for that
```

But in WebGL, texture UVs, start from bottom left in screen space (first vertex in buffer). Please ensure this is desired behaviour for the spec. If `gl_FragCoord` is used or similar approach to get screen space coords, then they will need to be Y flipped to match depth data. This is not aligned with WebGL, and worth checking if WebGPU uses bottom-left or top-left corner as 0,0 for UVs in order to ensure consistency for developers.

> Can you check if things work for you on our test page (note that texture upload uses Uint8Array because WebGL1 mandates that)?

Uint8Array is used, because it is 8 bits per channel, not 16 bits per channel. And in test provided Uint8Array is used as well, which is correct. So either test should be changed to Uint16Array (which I do not see possible, as there is no such WebGL format), or specs should change for that.

Reply:
Ah, correct, the prototype implementation used more WebGL-friendly convention for the returned data buffer, but I have received feedback that it's better to make indexing into the texture on the CPU simpler by putting the origin of the depth data in the top-left corner with Xs growing to the right and Ys growing downward - this makes it easier when interfacing with gamepad objects and possibly other events (and happens to follow the canvas2d convention). Unfortunately, it makes the WebGL case a bit trickier. This is mentioned in the spec by introducing something we call [normalized view coordinates](https://immersive-web.github.io/depth-sensing/#normalized-view-coordinates), I tried to make it explicit by adding a [note](https://immersive-web.github.io/depth-sensing/#ref-for-normalized-view-coordinates) in the spec as well.

As for Uint16Array, I think I am explaining this incorrectly. :( When you want to upload the data to the GPU, the Uint8Array is definitely the way to go, precisely because of what you are describing (note that the test page's shader [reassembles](https://source.chromium.org/chromium/chromium/src/+/master:third_party/webxr_test_pages/webxr-samples/shaders/depth-api-gpu.frag;l=15) 2 channels into an uint16 value because we do not have other means of sending 16-bit values to the GPU on WebGL1, so we had to pick a format that has 2 bytes per pixel and work with that). But when wanting to access the data on the CPU directly via the buffer, it is easier to use Uint16Array - this is what the explainer attempts to demonstrate here. If we wanted to use Uint8Array to get the value on the CPU, we would have to do something like:
```
const index = c + r * depthInfo.width;
// Depth API's "luminance-alpha" format is 2 bytes per pixel, so we need to reassemble the value.
// The below assumes little-endian architectures and should be equivalent to the GPU sample,
// specifically to dot(packedDepthAndVisibility, vec2(255.0, 256.0 * 255.0)) * uRawValueToMeters.
const depthInMetres = (uint8Data[2 * index] + 255 * uint8Data[2 * index + 1]) * depthInfo.rawValueToMeters;
```
By using Uint16Array as a window into the raw depth buffer, we can simplify indexing into the array and we don't need to worry about reassembling the values - Uint16Array does this for us.

Reply:
> Ah, correct, the prototype implementation used more WebGL-friendly convention for the returned data buffer, but I have received feedback that it's better to make indexing into the texture on the CPU simpler by putting the origin of the depth data in the top-left corner with Xs growing to the right and Ys growing downward - this makes it easier when interfacing with gamepad objects and possibly other events (and happens to follow the canvas2d convention). Unfortunately, it makes the WebGL case a bit trickier. This is mentioned in the spec by introducing something we call [normalized view coordinates](https://immersive-web.github.io/depth-sensing/#normalized-view-coordinates), I tried to make it explicit by adding a [note](https://immersive-web.github.io/depth-sensing/#ref-for-normalized-view-coordinates) in the spec as well.

WebGPU I believe is using the top-left origin as well, which makes it easier for the future. I believe docs should be a bit more explicit on this. And if Y-down is the preferred option, it perhaps should be consistent between CPU and GPU paths, so as to avoid confusions.

Regarding Uint16Array, with the current API design which has CPU and GPU paths, it will be still common for applications to use both paths, the difference is that it is chosen preferred method and then implements the other part on a less efficient path. I believe it is not a common pattern in API design for Web.

So here is the case:
1. Developer chooses to prefer GPU more than CPU, but cannot guarantee that platform supports GPU (by the spec).
2. If GPU path is provided, then CPU based functions will access texture buffer to read pixel and provide CPU access to depth information.
3. If CPU path is provided, then will re-create GPU texture in order to provide consistent functionality as with GPU path.

So that developer has full coverage: CPU and GPU, not **or**.  
CPU path data is intended for CPU usage, but should not make it harder to provide the ability to implement GPU path by the developer. That is why I'm a bit confused, docs mention LUMINANCE_ALPHA, which is against the current WebGL API means 2 bytes per pixel (8 bits per channel), which is Uint8Array for WebGL. But then CPU access is different and uses Uint16Array view (which is fine), but spec should be clear on that probably. In order to assist developers in implementing both paths based on the provided usage path.

Also, there should be no difference between origin of data, regardless of CPU or GPU paths, otherwise buffer/shader should be different depending on CPU or GPU path, which is not obvious for developer on why. Consistent Y-down origin across data - is better than complexity on describing when it is Y-down or Y-up.

Reply:
Thanks, I'll try to clarify the explainer!

> (...) it will be still common for applications to use both paths (...)

Do you think the applications will mostly interact with the raw array buffer on the CPU still, or will they rely more on the `depthInfo.getDeptInMeters(x, y)` helper method?

> Also, there should be no difference between origin of data, regardless of CPU or GPU paths (...)

100% agree! Do you think that currently there is a mismatch in where the origins are depending on where the data originated? That's definitely not the intent here! The assumption is that you always start from normalized view coordinates to index into the buffer, and it should not matter where it originated (if there are differences in the coordinate system origins, they will be hidden in `normDepthBufferFromNormView` transform that the app always needs to remember to apply).

Reply:
Sorry for the long delay. I've implemented the latest changes based on present specs: https://github.com/playcanvas/engine/pull/2947

And here are the findings:
1. Typos in explainer, still relevant, and proposed fix in this PR is relevant.
2. In order to construct texture from CPU L8A8 path, we have to construct buffer like so: `new Uint8Array(depthInfoCpu.data);`



> Do you think the applications will mostly interact with the raw array buffer on the CPU still, or will they rely more on the `depthInfo.getDeptInMeters(x, y)` helper method?

I believe they will rely on this method, yes. But explainer needs updating too (the apendix), as it currently not in sync with the latest changes to that method. It uses UV instead of XY (column, row).



> 100% agree! Do you think that currently there is a mismatch in where the origins are depending on where the data originated? That's definitely not the intent here! The assumption is that you always start from normalized view coordinates to index into the buffer, and it should not matter where it originated (if there are differences in the coordinate system origins, they will be hidden in `normDepthBufferFromNormView` transform that the app always needs to remember to apply).

I have not experienced different origins, as were only able to test CPU path, as GPU path on Chrome Android for me is not available yet.

It would be great for Specs to be more clear on origins for the data, is it top-left or bottom-left corner origin, as it is constantly relevant in GAPIs. Also, it was a change since previous specs.